xds/src/main/java/io/grpc/xds/XdsLrsClient.java

```diff
@@ -0,0 +1,325 @@
+  class LrsRpcRetryTask implements Runnable {
+
+    @Override
+    public void run() {
+      startLrsRpc();
+    }
+  }
+
+  // TODO (chengyuanzhang): consider putting to a util class.
+  @VisibleForTesting
+  static Duration numToDuration(long duration, TimeUnit unit) {
+    long time = unit.toNanos(duration);
+    long sec = time / 1_000_000_000L;
+    int nanos = (int) (time % 1_000_000_000L);
+    return Duration.newBuilder().setSeconds(sec).setNanos(nanos).build();
+  }
+
+  private class LrsStream implements StreamObserver<LoadStatsResponse> {
+
+    final LoadReportingServiceGrpc.LoadReportingServiceStub stub;
+    final Stopwatch stopwatch;
+    StreamObserver<LoadStatsRequest> lrsRequestWriter;
+    boolean initialResponseReceived;
+    boolean closed;
+    long loadReportIntervalNano = -1;
+    ScheduledHandle loadReportTimer;
+
+    LrsStream(LoadReportingServiceGrpc.LoadReportingServiceStub stub, Stopwatch stopwatch) {
+      this.stub = checkNotNull(stub, "stub");
+      this.stopwatch = checkNotNull(stopwatch, "stopwatch");
+    }
+
+    void start() {
+      lrsRequestWriter = stub.withWaitForReady().streamLoadStats(this);
+    }
+
+    @Override
+    public void onNext(final LoadStatsResponse response) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleResponse(response);
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(Status.fromThrowable(t)
+              .augmentDescription("Stream to XDS management server had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("Stream to XDS management server was closed"));
+        }
+      });
+    }
+
+    private void sendLoadReport() {
+      long interval = stopwatch.elapsed(TimeUnit.NANOSECONDS);
+      stopwatch.reset().start();
+      // TODO: get load report and send to management server
+      lrsRequestWriter.onNext(LoadStatsRequest.newBuilder()
+          .setNode(Node.newBuilder()
+              .setMetadata(Struct.newBuilder()
+                  .putFields(
+                      TRAFFICDIRECTOR_GRPC_HOSTNAME,
+                      Value.newBuilder().setStringValue(serviceName).build())))
+          .addClusterStats(ClusterStats.newBuilder().setLoadReportInterval(
+              numToDuration(interval, TimeUnit.NANOSECONDS)))
+          .build());
+      scheduleNextLoadReport();
+    }
+
+    private void scheduleNextLoadReport() {
+      if (loadReportIntervalNano > 0) {
+        // Cancel pending load report and reschedule with updated load reporting interval.
+        if (loadReportTimer != null && loadReportTimer.isPending()) {
```

Comment:
You probably want to move the cancellation out of the `if (loadReportIntervalNano > 0)`, because you'd cancel it no matter what.

Reply:
ACK.